core/device_impl.cpp

```diff
@@ -440,6 +441,105 @@ void DeviceImpl::get_param_float_async(const std::string &name,
                                             callback));
 }
 
+MavlinkCommands::Result DeviceImpl::set_flight_mode(FlightMode device_mode)
+{
+    uint8_t flag_safety_armed = is_armed() ? MAV_MODE_FLAG_SAFETY_ARMED : 0;
+
+    uint8_t mode = MAV_MODE_FLAG_CUSTOM_MODE_ENABLED | flag_safety_armed;
+
+    // Note: the safety flag is not needed in future versions of the PX4 Firmware
+    //       but want to be rather safe than sorry.
+    uint8_t custom_mode = px4::PX4_CUSTOM_MAIN_MODE_AUTO;
+    uint8_t custom_sub_mode = 0;
+
+    switch (device_mode) {
+        case FlightMode::HOLD:
+            custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_LOITER;
+            break;
+        case FlightMode::RETURN_TO_LAUNCH:
+            custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_RTL;
+            break;
+        case FlightMode::TAKEOFF:
+            custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_TAKEOFF;
+            break;
+        case FlightMode::LAND:
+            custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_LAND;
+            break;
+        case FlightMode::MISSION:
+            custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_MISSION;
+            break;
+        case FlightMode::FOLLOW_ME:
+            custom_sub_mode = px4::PX4_CUSTOM_SUB_MODE_AUTO_FOLLOW_TARGET;
+            break;
+        case FlightMode::OFFBOARD:
+            custom_mode = px4::PX4_CUSTOM_MAIN_MODE_OFFBOARD;
+            break;
+        default :
```

Comment:
In this case it would be helpful to add a `LogErr() << "Unknown Flight mode chosen.";`